src/runtime/internal/block_allocator.h

```diff
@@ -0,0 +1,477 @@
+
+    // Locates the "best-fit" block entry for the requested size, or nullptr if none was found
+    BlockEntry *find_block_entry(void *user_context, const MemoryProperties &properties, size_t size, bool dedicated);
+
+    // Creates a new block entry and int the list
+    BlockEntry *create_block_entry(void *user_context, const MemoryProperties &properties, size_t size, bool dedicated);
+
+    // Releases the block entry from being used, and makes it available for further allocations
+    void release_block_entry(void *user_context, BlockEntry *block_entry);
+
+    // Destroys the block entry and removes it from the list
+    void destroy_block_entry(void *user_context, BlockEntry *block_entry);
+
+    // Invokes the allocation callback to allocate memory for the block region
+    void alloc_memory_block(void *user_context, BlockResource *block);
+
+    // Invokes the deallocation callback to free memory for the memory block
+    void free_memory_block(void *user_context, BlockResource *block);
+
+    // Returns a constrained size for the requested size based on config parameters
+    size_t constrain_requested_size(size_t size) const;
+
+    // Returns true if the given block is compatible with the given properties
+    bool is_compatible_block(const BlockResource *block, const MemoryProperties &properties) const;
+
+    Config config;
+    LinkedList block_list;
+    MemoryAllocators allocators;
+};
+
+BlockAllocator *BlockAllocator::create(void *user_context, const Config &cfg, const MemoryAllocators &allocators) {
+    halide_abort_if_false(user_context, allocators.system.allocate != nullptr);
+    BlockAllocator *result = reinterpret_cast<BlockAllocator *>(
+        allocators.system.allocate(user_context, sizeof(BlockAllocator)));
+
+    if (result == nullptr) {
+        error(user_context) << "BlockAllocator: Failed to create instance! Out of memory!\n";
+        return nullptr;
+    }
+
+    result->initialize(user_context, cfg, allocators);
+    return result;
+}
+
+void BlockAllocator::destroy(void *user_context, BlockAllocator *instance) {
+    halide_abort_if_false(user_context, instance != nullptr);
+    const MemoryAllocators &allocators = instance->allocators;
+    instance->destroy(user_context);
+    halide_abort_if_false(user_context, allocators.system.deallocate != nullptr);
+    allocators.system.deallocate(user_context, instance);
+}
+
+void BlockAllocator::initialize(void *user_context, const Config &cfg, const MemoryAllocators &ma) {
+    config = cfg;
+    allocators = ma;
+    block_list.initialize(user_context,
+                          sizeof(BlockResource),
+                          config.initial_capacity,
+                          allocators.system);
+}
+
+MemoryRegion *BlockAllocator::reserve(void *user_context, const MemoryRequest &request) {
+
+    debug(nullptr) << "BlockAllocator: Reserve ("
```

Comment:
Here and elsewhere: when we actually have `user_context` available (as we do here), `debug(user_context)` is what you should do; `debug(nullptr)` is only for places that no user_context could possibly be available, e.g. static ctors or dtors

Reply:
Good catch!  Thanks ... I'll fix these occurrences!